In [1]:
import numpy as np

# Subway ridership for 5 stations on 10 different days
ridership = np.array([
    [   0,    0,    2,    5,    0],
    [1478, 3877, 3674, 2328, 2539],
    [1613, 4088, 3991, 6461, 2691],
    [1560, 3392, 3826, 4787, 2613],
    [1608, 4802, 3932, 4477, 2705],
    [1576, 3933, 3909, 4979, 2685],
    [  95,  229,  255,  496,  201],
    [   2,    0,    1,   27,    0],
    [1438, 3785, 3589, 4174, 2215],
    [1342, 4043, 4009, 4665, 3033]
])


In [6]:

# Change False to True for each block of code to see what it does

# Accessing elements
if True:
    print (ridership[1, 3])
    print (ridership[1:3, 3:5])
    print (ridership[1, :])
    

2328
[[2328 2539]
 [6461 2691]]
[1478 3877 3674 2328 2539]


In [9]:
# Vectorized operations on rows or columns
if True:
    print (ridership[0, :] + ridership[1, :])
    print (ridership[:, 0] + ridership[:, 1])
    


[1478 3877 3676 2333 2539]
[   0 5355 5701 4952 6410 5509  324    2 5223 5385]


In [10]:
# Vectorized operations on entire arrays
if True:
    a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    b = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])
    print (a + b)


[[ 2  3  4]
 [ 6  7  8]
 [10 11 12]]


In [17]:

def mean_riders_for_max_station(ridership):
    print (ridership.argmax[0:1])
    print (ridership.mean[0:1])
    print (ridership.mean())
    
    '''
    Fill in this function to find the station with the maximum riders on the
    first day, then return the mean riders per day for that station. Also
    return the mean ridership overall for comparsion.
    
    Hint: NumPy's argmax() function might be useful:
    http://docs.scipy.org/doc/numpy/reference/generated/numpy.argmax.html
    '''
    overall_mean = None # Replace this with your code
    mean_for_max = None # Replace this with your code
    
    return (overall_mean, mean_for_max)

In [37]:
# Find the station with the maximum riders on the first day
def mean_riders_for_max_station(ridership):
#ridership[0,:].argmax()
    max_station = ridership[0, :].argmax()

# Find the mean riders per day for the station
#ridership[:, max_station] 
    mean_for_max = ridership[:, max_station].mean()

# Find the mean ridership overall for comparison
    overall_mean = ridership.mean()

    return (overall_mean, mean_for_max)

In [40]:
mean_riders_for_max_station(ridership)

#overall mean = 2342.599

(2342.5999999999999, 3239.9000000000001)

## Overall mean

2342.5999999999999

## Mean for Max Station
3239.9000000000001


The station with the highest first day average had the highest overall average.



# Numpy Axis

### Operations along an axis




In [43]:

    # Subway ridership for 5 stations on 10 different days
ridership = np.array([
    [   0,    0,    2,    5,    0],
    [1478, 3877, 3674, 2328, 2539],
    [1613, 4088, 3991, 6461, 2691],
    [1560, 3392, 3826, 4787, 2613],
    [1608, 4802, 3932, 4477, 2705],
    [1576, 3933, 3909, 4979, 2685],
    [  95,  229,  255,  496,  201],
    [   2,    0,    1,   27,    0],
    [1438, 3785, 3589, 4174, 2215],
    [1342, 4043, 4009, 4665, 3033]
])

    
    [0, 3877, 4088, 3392, 4802, 3933,  229,    0, 3785, 4043]
    [2, 3674, 3991, 3826, 3932, 3909,  255,    1, 3589, 4009]
    [5, 2328, 6461, 4787, 4477, 4979,  496,   27, 4174, 4665]
    [0, 2539, 2691, 2613, 2705, 2685,  201,    0, 2215, 3033]

In [77]:
ridership[:, 4]

array([   0, 2539, 2691, 2613, 2705, 2685,  201,    0, 2215, 3033])

In [102]:
def mean_and_max_riders_per_day(ridership):
# Find the mean ridership per day for each subway station

#ridership.mean(axis=0)

# Will return 5 means, 1 for each subway station.
    station_riders = ridership.mean(axis=0)

# Find the maximum daily ridership and the minimum

    max_daily_ridership = station_riders.max()
    min_daily_ridership = station_riders.min()
    
    return (max_daily_ridership, min_daily_ridership) 

In [103]:
mean_and_max_riders_per_day(ridership)

(3239.9000000000001, 1071.2)

# Make my own Table

In [61]:
import pandas as pd

enrollments_df = pd.DataFrame({
        'account_key': [448, 448, 448, 448, 448],
        'status': ['canceled', 'canceled', 'canceled', 'canceled','current'],
        'join_date': ['2014-11-10', '2014-11-05', '2015-01-27', '2014-11-10', '2014-11-09'],
        'days_to_cancel': [65, 5, 0, 0, np.nan],
        'is_udacity': [True, True, True, True, True]
})

In [62]:
enrollments_df

account_key  days_to_cancel is_udacity   join_date    status
0          448            65.0       True  2014-11-10  canceled
1          448             5.0       True  2014-11-05  canceled
2          448             0.0       True  2015-01-27  canceled
3          448             0.0       True  2014-11-10  canceled
4          448             NaN       True  2014-11-09   current

In [63]:
enrollments_df.mean()

account_key       448.0
days_to_cancel     17.5
is_udacity          1.0
dtype: float64

In [83]:
ridership_df = pd.DataFrame({
    'R003': [0, 1478, 1613, 1560, 1608, 1576,   95,    2, 1438, 1342],
    'R004': [0, 3877, 4088, 3392, 4802, 3933,  229,    0, 3785, 4043],
    'R005': [2, 3674, 3991, 3826, 3932, 3909,  255,    1, 3589, 4009],
    'R006': [5, 2328, 6461, 4787, 4477, 4979,  496,   27, 4174, 4665],
    'R007': [0, 2539, 2691, 2613, 2705, 2685,  201,    0, 2215, 3033]
}, index=[
    '05-01-11', '05-02-11', '05-03-11', '05-04-11', '05-05-11',
    '05-06-11', '05-07-11', '05-08-11', '05-09-11', '05-10-11'    
])    



In [84]:
ridership_df

R003  R004  R005  R006  R007
05-01-11     0     0     2     5     0
05-02-11  1478  3877  3674  2328  2539
05-03-11  1613  4088  3991  6461  2691
05-04-11  1560  3392  3826  4787  2613
05-05-11  1608  4802  3932  4477  2705
05-06-11  1576  3933  3909  4979  2685
05-07-11    95   229   255   496   201
05-08-11     2     0     1    27     0
05-09-11  1438  3785  3589  4174  2215
05-10-11  1342  4043  4009  4665  3033

In [86]:
ridership_df.loc['05-01-11']

R003    0
R004    0
R005    2
R006    5
R007    0
Name: 05-01-11, dtype: int64

In [88]:
ridership_df.iloc[3]
#access by position

R003    1560
R004    3392
R005    3826
R006    4787
R007    2613
Name: 05-04-11, dtype: int64

In [89]:
ridership_df.iloc[0, 3]
#gives position of row 0, column 3 "R006"

5

In [91]:
ridership_df.loc['05-04-11', 'R004']

3392

In [96]:
ridership_df['R006']

05-01-11       5
05-02-11    2328
05-03-11    6461
05-04-11    4787
05-05-11    4477
05-06-11    4979
05-07-11     496
05-08-11      27
05-09-11    4174
05-10-11    4665
Name: R006, dtype: int64

In [97]:
ridership_df.values

array([[   0,    0,    2,    5,    0],
       [1478, 3877, 3674, 2328, 2539],
       [1613, 4088, 3991, 6461, 2691],
       [1560, 3392, 3826, 4787, 2613],
       [1608, 4802, 3932, 4477, 2705],
       [1576, 3933, 3909, 4979, 2685],
       [  95,  229,  255,  496,  201],
       [   2,    0,    1,   27,    0],
       [1438, 3785, 3589, 4174, 2215],
       [1342, 4043, 4009, 4665, 3033]], dtype=int64)

In [98]:
ridership_df.values.mean()

2342.5999999999999

In [108]:
import pandas as pd

# Subway ridership for 5 stations on 10 different days
ridership_df = pd.DataFrame(
    data=[[   0,    0,    2,    5,    0],
          [1478, 3877, 3674, 2328, 2539],
          [1613, 4088, 3991, 6461, 2691],
          [1560, 3392, 3826, 4787, 2613],
          [1608, 4802, 3932, 4477, 2705],
          [1576, 3933, 3909, 4979, 2685],
          [  95,  229,  255,  496,  201],
          [   2,    0,    1,   27,    0],
          [1438, 3785, 3589, 4174, 2215],
          [1342, 4043, 4009, 4665, 3033]],
    index=['05-01-11', '05-02-11', '05-03-11', '05-04-11', '05-05-11',
           '05-06-11', '05-07-11', '05-08-11', '05-09-11', '05-10-11'],
    columns=['R003', 'R004', 'R005', 'R006', 'R007']
)

   

ridership_df


R003  R004  R005  R006  R007
05-01-11     0     0     2     5     0
05-02-11  1478  3877  3674  2328  2539
05-03-11  1613  4088  3991  6461  2691
05-04-11  1560  3392  3826  4787  2613
05-05-11  1608  4802  3932  4477  2705
05-06-11  1576  3933  3909  4979  2685
05-07-11    95   229   255   496   201
05-08-11     2     0     1    27     0
05-09-11  1438  3785  3589  4174  2215
05-10-11  1342  4043  4009  4665  3033

In [142]:
max_station = ridership_df.iloc[0].argmax()


In [139]:
mean_for_max = ridership[max_station].mean()

In [141]:
overall_mean = ridership_df.values.mean()

In [144]:
def mean_riders_for_max_station(ridership):
    '''
    Fill in this function to find the station with the maximum riders on the
    first day, then return the mean riders per day for that station. Also
    return the mean ridership overall for comparsion.
    
    This is the same as a previous exercise, but this time the
    input is a Pandas DataFrame rather than a 2D NumPy array.
    '''
    
    max_station = ridership_df.iloc[0].argmax()
    mean_for_max = ridership[max_station].mean()
    overall_mean = ridership_df.values.mean()

    
    
    return (overall_mean, mean_for_max)

In [146]:
mean_riders_for_max_station(ridership_df)

(2342.5999999999999, 3239.9)